text/0110-periodic-full-compaction.md

```diff
@@ -0,0 +1,386 @@
+# Periodic Full Compaction
+
+Author: [Alex Feinberg](https://github.com/afeinberg)
+
+Reviewers: [Connor](https://github.com/Connor1996),
+[Tony](https://github.com/tonyxuqqi), [Andy](https://github.com/v01dstar),
+others
+
+## Introduction
+
+**Periodic full compaction** is a scheduled task that starts at specified times
+of the day on each TiKV node and compacts the column families of all regions *at
+all levels including the bottommost (L6)*. In order to reduce to impact running
+such a heavy-weight task would have on the cluster's ability to serve traffic,
+full compaction is incremental: before the next range (presently a region) is
+compacted, we check if the load is below a certain threshold; if the threshold
+is exceeded, we pause the task until the load is again below a certain
+threshold.
+
+## Motivation
+
+Presently, we have no way to periodically execute a RocksDB compaction across
+all levels. This has a number of implications: compaction filters (used to
+remove tombstones) only run when the bottom-most (L6) compaction is executed; a
+system with a high number of deletes that experiences a heavy read-only (but
+little or no writes) might thus have accumulated tombstones markers that are not
+deleted.
+
+Using `tikv-ctl compact-cluster` is not suitable for this goal: while it executes
+a full compaction, it may impact online user traffic which makes it non-viable
+for production usage without downtime. Periodic full compaction provides a way to
+achieve the same goal in a more controllable way.
+
+## Detailed design
+
+### Periodic scheduling
+
+#### Configuration
+
+To enable periodic full compaction, specify the hours during which we wish to
+schedule full compaction to run in tikv's configuration and a maximum CPU
+utilization threshold. CPU utilization is calculated by using process stats in
+`/proc` over a 10-minute window. (See *Conditions for full compaction to run*
+below.)
+
+>
+> `tikv.toml` setting to run compaction at 03:00 and 23:00
+> (3am and 11pm respectively) in the tikv nodes' local timezone if CPU
+> usage is below 90%:
+>
+> ```toml
+>[raftstore]
+>periodic-full-compact-start-max-cpu = 0.9
```

Comment:
Do you need to consider the compact of the dist?

Reply:
@bufferflies  What do you mean by `compact of the dist`?

Reply:
I mean that the impact of dist.

Reply:
I believe, no, this is independent and runs on each tikv node (if by dist you mean the distributed aspect.) If you mean the key distribution, it is also independent of this.

Reply:
In my understant, the compact may bring extra disk IO, so it impact the write duration of  other component like wal or flush.

Reply:
Oh,you typed 'dist' instead of 'disk', which confused me. Yes, there is impact on all kinds of live traffic (including on disk), but it can be mitigated by configuring full compaction to run at times when it is least loaded.

In the future (see future work), it will be possible to use other metrics with full compaction, e.g., pausing when disk usage is high.